lib/fluent/plugin/in_tail.rb

```diff
@@ -370,17 +370,19 @@ def existence_path
     def refresh_watchers
       target_paths_hash = expand_paths
       existence_paths_hash = existence_path
-      
+
       log.debug {
         target_paths_str = target_paths_hash.collect { |key, target_info| target_info.path }.join(",")
         existence_paths_str = existence_paths_hash.collect { |key, target_info| target_info.path }.join(",")
         "tailing paths: target = #{target_paths_str} | existing = #{existence_paths_str}"
       }
 
-      unwatched_hash = existence_paths_hash.reject {|key, value| target_paths_hash.key?(key)}
+      @pf&.unwatch_removed_targets(target_paths_hash)
```

Comment:
Although I might overlook something, but could you let me ask the following?

I considered the pattern that /var/log/test1.log was removed then its inode(inode-1) was reused by /var/log/test2.log in the setting `path = /var/log/*.log*`.

- target_paths_hash will include inode-1, since /var/log/test2.log has the same inode as the removed /var/log/test1.log.
- So `@pf&.unwatch_removed_targets(target_paths_hash)` will skip inode-1 so the position info of inode-1 won't be reset.
- `removed_hash` also won't include inode-1 due to the same reason.

In this case, won't fluentd misrecognize the previous position for /var/log/test2.log?

Reply:
> That scenario can happen in the current master branch, right?

I wanted to say that https://github.com/fluent/fluentd/pull/4237#discussion_r1260712488 is the side effect introduced by adding `@pf&.unwatch_removed_targets(target_paths_hash)` in your patch.

But maybe I was wrong, because `added_hash` also won't include inode-1 in my suggested pattern.
I rewrote the scenario a bit as follows.

In the pattern that `/var/log/test1.log` was removed then its inode(inode-1) was reused by `/var/log/test2.log` in the setting `path = /var/log/*.log*`:
- `target_paths_hash` will include inode-1, since `/var/log/test2.log` has the same inode as the removed `/var/log/test1.log`.
  - It means that `@pf&.unwatch_removed_targets(target_paths_hash)` will skip inode-1 so the position info of inode-1 won't be reset.
  - `removed_hash` also won't include inode-1 due to the same reason.
  - `added_hash` also won't include inode-1 due to the same reason.
- So `refresh_watchers()` will call neither `start_watchers()` nor `stop_watchers()` for both of `/var/log/test1.log` and `/var/log/test2.log`.

Yes, the above scenario can happen even in the current master branch.

The situation will be resolved after a new `/var/log/test1.log` was created and `@tails["/var/log/test1.log"]` got updated with a new inode.
So it won't be a big problem in real environment, I believe...

Reply:
Is it rotation? (Is the `test1.log` the same file as `test2.log`?)

If so, it is correct that `refresh_watcher` does nothing for `inode-1` because it means the target `inode-1` still exists.
In this case, `update_watcher` update the TailWatcher.

However, I wonder the old TailWathcer can collect duplicate log in `rotate_wait` interval...
I'll check it.

Reply:
> However, I wonder the old TailWathcer can collect duplicate log in rotate_wait interval...
I'll check it.

This may not be the point that you were originally concerned about, but I found this can cause log duplication.
This seems to be a problem that has existed for a long time.

This is a similar problem to https://github.com/fluent/fluentd/pull/4237#issuecomment-1632176184
It seems that in_tail has a log duplication problem since a long time ago with `rotate_wait`.

I will explain this in another comment.

Reply:
I commented on this in https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632